### PR TITLE
[SPARK-20684][R] Expose create[OrReplace]GlobalTempView and dropGlobalTempView in SparkR

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

This PR aims to expose `createGlobalTempView` and `dropGlobalTempView` in SparkR.

## How was this patch tested?

Pass the Jenkins.

### CODE REVIEW

File: R/pkg/R/DataFrame.R [hunk 1]
```diff
@@ -501,6 +501,34 @@ setMethod("createOrReplaceTempView",
               invisible(callJMethod(x@sdf, "createOrReplaceTempView", viewName))
           })
 
+#' Creates a global temporary view using the given name.
```
Comment (falaki) [inline R/pkg/R/DataFrame.R]: Would you please extend this documentation to explain what is the difference between a global temp view and a normal temp view? I constantly hear from SparkR users that our documentation simply lists the function arguments.

File: R/pkg/inst/tests/testthat/test_sparkSQL.R [hunk 2]
```diff
@@ -721,6 +721,16 @@ test_that(
...
 })
 
+test_that(
+  "createGlobalTempView() results in a queryable table and sql() results in a new DataFrame", {
+  df <- read.json(jsonPath)
+  createGlobalTempView(df, "global_table1")
+  newdf <- sql("SELECT * FROM global_temp.global_table1 where name = 'Michael'")
+  expect_is(newdf, "SparkDataFrame")
+  expect_equal(count(newdf), 1)
+  expect_true(dropGlobalTempView("global_table1"))
```
Comment (falaki) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: Please verify that the view is indeed dropped by comparing result of `tableNames()` after the drop call to what it was at the beginning.

**Review by falaki — CHANGES_REQUESTED** [review]
Thanks @dongjoon-hyun I left two comments.

File: R/pkg/inst/tests/testthat/test_sparkSQL.R (hunk 2 above)
Comment (dongjoon-hyun) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: Sure!

Comment (dongjoon-hyun) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: Ur, if you don't mind, I will use the way to count tables like 
https://github.com/apache/spark/commit/b952b44af4d243f1e3ad88bccf4af7d04df3fc81 .

Comment (dongjoon-hyun) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: I found that global temp tables are now returned by `listTables`. (R is the same).
```
scala> spark.range(10).createGlobalTempView("t1")

scala> spark.catalog.tableExists("global_temp.t1")
res8: Boolean = true

scala> spark.catalog.listTables().show
+----+--------+-----------+---------+-----------+
|name|database|description|tableType|isTemporary|
+----+--------+-----------+---------+-----------+
+----+--------+-----------+---------+-----------+
```

File: R/pkg/R/DataFrame.R (hunk 1 above)
Comment (dongjoon-hyun) [inline R/pkg/R/DataFrame.R]: I updated the doc.

File: R/pkg/inst/tests/testthat/test_sparkSQL.R [hunk 3]
```diff
@@ -721,6 +721,17 @@ test_that(
...
+  expect_false(dropGlobalTempView("global_table1"))
```
Comment (dongjoon-hyun) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: Instead, I added the above line to check the success of deletion in the previous line. How do you think about this?

**Comment by felixcheung** [comment]
actually, it was intentional that we didn't add it in R since there's always one session only.
Could you elaborate how global view helps with cross language?

**Comment by dongjoon-hyun** [comment]
Oh, I assumed there exists a case about cross languages inside SparkR and thought this is another API parity issue at first. Maybe, @falaki can give us the goal described in JIRA.

**Comment by felixcheung** [comment]
FYI, this was SPARK-17865 and decided not needed.

**Comment by dongjoon-hyun** [comment]
Yep. It looks like that. At that time, @rxin was the reporter. And this time, @falaki is the reporter.

@felixcheung . Zeppelin can share the Spark context between Java and Scala, can't it?

Can Databricks cloud share the context (or JVM) among more languages? If then, that would be great.

**Comment by dongjoon-hyun** [comment]
Thank you for comments, @falaki and @felixcheung . I added the duplication link to the issue, SPARK-20684, and ask @falaki to close the JIRA issue because he is the reporter.

**Comment by falaki** [comment]
@dongjoon-hyun and @felixcheung just to confirm, yes we do share SparkContext across languages in Databricks.

I think this is a useful API in general. Note that when user creates spark session in SparkR if there is an existing `SQLContext` that is returned.

**Comment by dongjoon-hyun** [comment]
Thank you for confirming. If then, what about merging this for both API parity in Apache Spark and preparing the furture?

**Comment by dongjoon-hyun** [comment]
Also, ping @shivaram and @yanboliang .

**Comment by felixcheung** [comment]
there are several different ways to share context or `SparkSession` between languages and I think that discussion is orthogonal to the global session.

is there value having temp view in global session? do temp views in global session disappear when the SparkContext is stopped? that was my understanding earlier and I offered in SPARK-17865 several different approaches where the global session would make sense in that case. otherwise since SparkR itself shutdown both SparkSession and SparkContext together, I think it will be very confusing to introduce an API that "does not work" in the SparkR package.

**Comment by falaki** [comment]
@felixcheung we all know that SparkR (and in general R) API is not perfect when it comes to ETLing unstructured data. For example we don't have a great story for nested data, etc. To overcome these limitations many ETL their data in Python or Scala and then analyze them in R.

With introduction of sessions that workflow is partially broken. You can still do it but you need to persist the table. The global temp view is to solve that problem. It exists in PySpark, so I think it deserves to exist in SparkR as well.

**Comment by rxin** [comment]
@felixcheung what's your concern with this one? seems like just for api parity sake we should add this?

**Comment by felixcheung** [comment]
@rxin it was from a while back in SPARK-17865
essentially these are exactly the same in standalone SparkR:
```
sparkR.session()
...
createGlobalTempView(df, "foo")
sparkR.session.stop()
```
same as
```
sparkR.session()
...
createOrReplaceTempView(df, "foo")
sparkR.session.stop()
```
this is because in SparkR, `SparkContext` is stopped the same time as `SparkSession` (and there is only one session at a time)
so I think it makes sense to formally support the idea of the "global session" as discussed in SPARK-17865; with some changes in session initialization and session termination, and R session cleanup, plus more documentation and proper testing of global view lifetime beyond session, it should be good to have.

**Comment by falaki** [comment]
@felixcheung How about before `sparkR.session.stop()`? The use case I am trying to recover (which is broken now) is sharing table in SparkR and working on it in a different language.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by cloud-fan** [comment]
`sparkR.session.stop()` is a problem, but I think this patch still makes sense to share global temp views between different sessions, e.g. between Spark R and PySpark. We should not block this PR because of it.

**Comment by cloud-fan** [comment]
retest this please

File: R/pkg/NAMESPACE [hunk 4]
```diff
@@ -98,6 +98,7 @@ exportMethods("arrange",
               "corr",
               "covar_samp",
               "covar_pop",
+              "createGlobalTempView",
```
Comment (cloud-fan) [inline R/pkg/NAMESPACE]: there is also a `createOrReplaceGlobalTempView`

Comment (dongjoon-hyun) [inline R/pkg/NAMESPACE]: Oh, I'll expose that too.

**Comment by felixcheung** [comment]
Given the timing are you suggesting we include this in 2.2.0?

This hasn't been review yet.

**Comment by dongjoon-hyun** [comment]
Could you review this, @felixcheung ?

**Comment by felixcheung** [comment]
@cloud-fan thank you for chiming in. I have been looking for some feedback and I'm hoping we could get something more definitive.

I'm sure there is a need to share context between different languages. Having been on the consuming side for a long time I'm surprised the lack of incentive to formalize how to share context, and to avoid private methods getting called, wire protocol reverse-engineered and so on.

I'm aware of at least 5 or 6 unique implementations of `R - to - Spark`. Among these, I'm pretty certain only `databricks` has the proper support for multiple SparkSessions. There is only 1 other implementation that has sharing and it does so at the SparkSession-level (ie. there is only one session across all languages).

So I think sharing context is an orthogonal question, and can be achieved in several different ways.

Even if there isn't multiple/default/active session support, or that it wouldn't applicable, I still think it makes sense to have the concept of a global session, if nothing else we could get that API parity. So all I'm saying is we should have that in SparkR so that any other APIs working with global sessions only would actually work with SparkR by itself.

And for that, a proposal was made, and from what I can see implementation isn't going to be hard.

What do people think?

**Comment by dongjoon-hyun** [comment]
If @felixcheung provides the real open source implementation for that in a near future, that would be perfect. Why don't you make a PR as a PoC first if the implementation isn't going to be hard?

**Comment by dongjoon-hyun** [comment]
Hi, @felixcheung .

IMO, this is just a confirmation on interface. This will only help future contributors (including you) make a PR on the implementation.

Or, if you want to include this PR in your PR, please make your PR first. I will take a look at that and close this happily.

**Comment by HyukjinKwon** [comment]
Just a question out of curiosity. Don't we have this in Python too? I just wonder if this support should be in a different way and/or held off for R only.  Is there any case it does not work incorrectly R-specifically?

File: R/pkg/R/DataFrame.R [hunk 5]
```diff
@@ -501,6 +501,72 @@ setMethod("createOrReplaceTempView",
...
+#'
+#' Creates a new global temporary view using a SparkDataFrame in the Spark Session.
+#' Global temporary view is cross-session while a temporary view is not. Its lifetime
+#' is the lifetime of the Spark application, i.e. it will be automatically dropped
+#' when the application terminates. It's tied to a system preserved database `global_temp`,
+#' and we must use the qualified name to refer a global temporary view, e.g.
+#' `SELECT * FROM global_temp.view1`. If a global temporary view with the same name
+#' already exists, it raises exception.
```
Comment (HyukjinKwon) [inline R/pkg/R/DataFrame.R]: I guess `` ` ... ` `` should be `\code{ ... }` if this is meant to be a code block.

BTW, I hope the documentations for the same API in Scala/Python/R are almost identical if these are not language-specific. This requires a bit of overhead when someone sweep them to fix the same contents and sometimes it needs several follow-up PRs.

**Comment by felixcheung** [comment]
Hyukjin appreciate your chiming in.

This is something that does not work for R (in SparkR standalone) specifically. For reasons and proposal to address please see the earlier comments.

**Comment by HyukjinKwon** [comment]
Ah, so, this API is useless in R alone but might be useful when it shares in other languages for the current status. The point is that this API has to be useful in R alone as well (and this should be primary) rather than accepting the current status. I see.

**Pull request closed without merging**